src/DynamoCoreWpf/ViewModels/Search/NodeAutoCompleteSearchViewModel.cs

```diff
@@ -53,42 +42,240 @@ internal void PopulateAutoCompleteCandidates()
         /// that output a type compatible with the port type if it's an input port.
         /// These search elements can belong to either zero touch, NodeModel or Builtin nodes.
         /// This method returns an empty collection if the input port type cannot be inferred or
-        /// there are no matching nodes found for the type. Currently the match is an exact match
-        /// done including the rank information in the type, e.g. Point[] or var[]..[].
-        /// The search elements can be made to appear in the node autocomplete search dialog.
+        /// there are no matching nodes found for the type. Currently the match does not include
+        /// rank information. For example Curve[] is matched as Curve. The results include types
+        /// that would be valid using normal class inheritance rules.
+        /// The resulting compatible search elements can be made to appear in the node autocomplete search dialog.
         /// </summary>
         /// <returns>collection of node search elements</returns>
         internal IEnumerable<NodeSearchElement> GetMatchingNodes()
         {
             var elements = new List<NodeSearchElement>();
-
             var inputPortType = PortViewModel.PortModel.GetInputPortType();
-            if (inputPortType == null) return elements;
-
-            var libraryServices = dynamoViewModel.Model.LibraryServices;
 
-            // Builtin functions and zero-touch functions
-            // Multi-return ports for these nodes do not contain type information
-            // and are therefore skipped.
-            var functionGroups = libraryServices.GetAllFunctionGroups();
-            var functionDescriptors = functionGroups.SelectMany(fg => fg.Functions).Where(fd => fd.IsVisibleInLibrary);
+            if (inputPortType == null) return elements;
+            //if inputPortType is an array, use just the typename
+            var cbn = ParserUtils.ParseWithCore($"dummyName:{ inputPortType};", dynamoViewModel.Model.LibraryServices.LibraryManagementCore);
+            var ast = cbn.CodeBlockNode.Children().FirstOrDefault() as IdentifierNode;
+            //if parsing the type failed, revert to original string.
+            inputPortType = ast != null ? ast.datatype.Name : inputPortType;
 
-            foreach (var descriptor in functionDescriptors)
+            //gather all ztsearchelements that are visible in search and filter using inputPortType and zt return type name.
+            var ztSearchElements = Model.SearchEntries.OfType<ZeroTouchSearchElement>().Where(x => x.IsVisibleInSearch);
+            foreach (var ztSearchElement in ztSearchElements)
             {
-                if (descriptor.ReturnType.ToString() == inputPortType)
+                //for now, remove rank from descriptors
```

Comment:
The reason is we using our own rank defined？

Reply:
please see this note in the description:

`when we encounter a type that includes a rank we strip off the rank by parsing the typedIdentifer and returning only the typeName which does not include the rank. This let's us match nodes that accept arrays with nodes that return single items and vice versa. This makes sense to me because of replication and how it is easy it is for a user to create a list of items from a node that returns single items.`

I'm not sure what you mean by "our own rank"

this PR is currently disregarding rank completely.

Reply:
Thanks for the description, understand better now!